src/v/rpc/transport.h

```diff
@@ -132,21 +132,47 @@ inline errc map_server_error(status status) {
 
 template<typename T>
 ss::future<result<rpc::client_context<T>>> parse_result(
-  ss::input_stream<char>& in, std::unique_ptr<streaming_context> sctx) {
+  ss::input_stream<char>& in,
+  std::unique_ptr<streaming_context> sctx,
+  transport_version req_ver) {
     using ret_t = result<rpc::client_context<T>>;
-    // check status first
-    auto st = static_cast<status>(sctx->get_header().meta);
-    if (st != status::success) {
-        /**
-         * signal that request body is parsed since it is empty when status
-         * indicates server error.
-         */
-        sctx->signal_body_parse();
 
+    const auto st = static_cast<status>(sctx->get_header().meta);
+    const auto rep_ver = sctx->get_header().version;
+
+    /*
+     * the reply version should always be the same as the request version,
+     * otherwise this is non-compliant behavior. the exception to this
+     * rule is a v0 reply to a v1 request (ie talking to old v0 server).
+     */
+    const auto protocol_violation
+      = rep_ver != req_ver
+        && (req_ver != transport_version::v1 || rep_ver != transport_version::v0);
+
+    if (unlikely(st != status::success || protocol_violation)) {
+        sctx->signal_body_parse();
+        if (st == status::version_not_supported) {
+            /*
+             * let version_not_supported take precedence over error handling for
+             * protocol violations because the protocol violation may be due to
+             * the unsupported version scenario.
+             */
+            return ss::make_ready_future<ret_t>(map_server_error(st));
+        }
+        if (protocol_violation) {
+            vlog(
+              rpclog.warn,
+              "Protocol violation: request version {} incompatible with "
+              "reply version {}",
+              req_ver,
+              rep_ver);
+        }
+        if (st == status::success) {
```

Comment:
Isn't this the only possible value of `st` if `protocol_violation` at this point? Would it be simpler if this scope was collapsed into the one above?

Reply:
i agree that if `protocol_violation` is `true` then it implies that `st == status::success` and could be folded into the previous scope.

but if `st != status::success` then `protocol_violation` could still be true and we'd still have to check `st`?

Reply:
I think Noah's, even if there is a protocol violation, if there's another error more specific than `errc::service_error` known to the streaming context, we probably should be using it instead. I guess the question is whether we can get here with an error in `sctx`, or whether that'd have been caught earlier

Reply:
Yep, I knew I was missing something.